Add unit tests for pkg/service.Service

brb opened the pull request:
This PR:

- Introduces the `lbmap.LBMap` interface and its mock implementation - `LBMockMap`.
- Adds `pkg/service.Service` unit tests.
- Fixes the bug where backends were not properly accounted which resulted in the backend leak.
- Removes the `lbmap.BackendAddrID` type.

Reviewable per commit.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9289)
<!-- Reviewable:end -->

brb committed:
lbmap: Add LBMap interface

The interface is going to be used to replace the actual implementation
of lbmap in the unit testing of pkg/service/service.go.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap: Add LBMockMap for testing

LBMockMap implements the LBMap interface, and can be used for testing.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
service: Fix missing tracking of backends by hash

The missing tracking resulted in backends not being removed from BPF
maps.

Fixes: a7a841e88c ("Add temporary package for service management")
Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
service: Add first batch of Service unit tests

Signed-off-by: Martynas Pumputis <m@lambda.lt>

joestringer commented on pkg/service/service.go:
```diff
@@ -62,6 +64,7 @@ func NewService() *Service {
 		svcByID:         map[lb.ID]*lb.LBSVC{},
 		backendRefCount: counter.StringCounter{},
 		backendByHash:   map[string]*lb.LBBackEnd{},
+		lbmap:           &lbmap.LBBPFMap{},
```
This should probably be passed in as a parameter to `NewService()`, as an interface rather than the concrete type?

brb replied:
Hmm, then we would need to pass the `&lbmap.LBBPFMap{}` in daemon.go, which I would like to avoid doing, as ideally daemon should not know anything about `lbmap`.

Alternatively, we could introduce a new constructor (`NewServiceWithLBMap(lbmap LBMap)`). But as long as we are testing `Service` from the same package, IMHO it's fine to modify its private field.

joestringer commented on pkg/maps/lbmap/types.go:
```diff
@@ -25,6 +25,16 @@ import (
 )
 
+// LBMap is the interface describing methods for manipulating service maps.
+type LBMap interface {
```
Shouldn't this interface be defined where it's used? Ie in `pkg/service`? Or are there dependency problems with `pkg/loadbalancer`?

brb replied:
I wasn't aware of the following recommendation:

> Go interfaces generally belong in the package that uses values of the interface type, not the package that implements those values. The implementing package should return concrete (usually pointer or struct) types: that way, new methods can be added to implementations without requiring extensive refactoring.

https://github.com/golang/go/wiki/CodeReviewComments#interfaces

Moved it to the `pkg/service`, as we currently have just one consumer of the interface.

joestringer reviewed (commented):
Couple of minor nits, but otherwise a great start. Thanks!

brb committed:
lbmap: Change DumpBackendMapsToUserspace() return type

Return list instead of map indexed with BackendAddrID which is no longer
used.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap: Get rid of unused BackendAddrID

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap: Add checks for NULL IDs

NULL IDs (neither for backend nor for service) aren't allowed. So,
return an error if a NULL ID is given when manipulating the lbmaps.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
service: Restore svc.BackendByHash upon svc restore

Otherwise, we will start leaking backends after cilium-agent has
restarted.

Fixes: a7a841e88c ("Add temporary package for service management")
Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
service: Add debug log statements

Should ease a debugging of the service management.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap: Remove unused DumpServiceMapsToUserspaceV2 return

This commit changes the DumpServiceMapsToUserspaceV2() declaration
by removing the unused return value SVCMap which values are already
in the returned SVCList.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap: Extend lbmap_mock

- Add DumpServiceMapsToUserspaceV2 and DumpBackendMapsToUserspace.
- Proper tracking of backends and services.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
service: Add unit test for restoring services

Signed-off-by: Martynas Pumputis <m@lambda.lt>

ianvernon commented on pkg/service/manager_test.go:
```diff
@@ -0,0 +1,111 @@
+// Copyright 2018 Authors of Cilium
```
2018 --> 2019

brb replied:
Oops, copy-pasta.
